### PR TITLE
Feature/reorder multi qubit gates

### PR DESCRIPTION
**Pull request by yao-cqc**

### CODE REVIEW

**Review by sjdilkes — CHANGES_REQUESTED** [review]
All readable and neat and well tested - just a few smallish things.

File: tket/src/Mapping/MultiGateReorder.cpp [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+  this->mapping_frontier_ =
+      std::make_shared<MappingFrontier>(*_mapping_frontier);
+  // This needs to be updated every time the frontier changes
+  this->u_frontier_edges_ =
+      convert_u_frontier_to_edges(*frontier_convert_vertport_to_edge(
+          _mapping_frontier->circuit_, _mapping_frontier->quantum_boundary));
+}
+
+// Traverse the DAG to find the UnitID associated with an edge
+Vertex get_input_from_vertex_edge(
```
[the remark this answers is not in the document]
Comment (yao-cqc) [reply tket/src/Mapping/MultiGateReorder.cpp:18]: I'm not sure how easy is it to update this information. Especially when gates are added during routing.

File: tket/src/Mapping/MultiGateReorder.cpp [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+void MultiGateReorder::solve() {
+  // Try to advance the frontier
+  this->mapping_frontier_->advance_frontier_boundary(this->architecture_);
+  this->u_frontier_edges_ =
+      convert_u_frontier_to_edges(*frontier_convert_vertport_to_edge(
+          this->mapping_frontier_->circuit_,
+          this->mapping_frontier_->quantum_boundary));
+  // Get a subcircuit only for iterating vertices
+  Subcircuit circ =
+      this->mapping_frontier_->get_frontier_subcircuit(UINT_MAX, UINT_MAX);
```
[the remark this answers is not in the document]
Comment (yao-cqc) [reply tket/src/Mapping/MultiGateReorder.cpp]: I added the parameters and set the default to UINT_MAXs. How do you feel about setting the default to UINT_MAX? The procedure isn't that expensive though.

[the remark this answers is not in the document]
Comment (sjdilkes) [reply tket/src/Mapping/MultiGateReorder.cpp]: The procedure isn't expensive, but it is likely being run many times through a full route, meaning making a subcircuit of the remaining Circuit each time will likely add up. I'd set the default lower for now (say 10) and we can work out a better set of default parameters later.

File: tket/src/Mapping/MultiGateReorder.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (sjdilkes) [reply tket/src/Mapping/MultiGateReorder.cpp:18]: Is it possible to at least traverse the DAG back to the MappingFrontier edges?

**Review by sjdilkes — CHANGES_REQUESTED** [review]
A continuation of prior conversation.

**Review by sjdilkes — CHANGES_REQUESTED** [review]
All looking good, though one final thing which I missed the other day - there are no  tests for MultiGateReorderRoutingMethod. I appreciate the functionality is tested already, but please could you add a simple couple of test cases for construction and calling methods, mainly for test coverage upkeep.

**Comment by yao-cqc** [comment]
I refactored some methods so they can be reused by`MultiGateReorderRoutingMethod::check_method`.

**Review by sjdilkes — APPROVED** [review]
Nice!

**Pull request merged**